File: app/src/main/java/hr/caellian/notestream/gui/fragments/FragmentBarPlayer.java
package hr.caellian.notestream.gui.fragments;

import android.app.Fragment;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import java.util.Timer;
import java.util.TimerTask;

import hr.caellian.notestream.NoteStream;
import hr.caellian.notestream.R;
import hr.caellian.notestream.gui.ActivityPlayer;
import hr.caellian.notestream.data.playable.Playable;
import hr.caellian.notestream.data.PlayerService;
import hr.caellian.notestream.util.RepeatState;

/**
 * Created by caellyan on 16/06/17.
 */

public class FragmentBarPlayer extends Fragment implements Playable.ControlListener, Playable.ProgressListener{

    View view;
    PlayerService.PlayerServiceBinder psb = NoteStream.getInstance().getPlayerServiceBinder();

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.bar_playing, container, false);

        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(view.getContext(), ActivityPlayer.class);
                intent.addFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT);
                startActivity(intent);
            }
        });

        final Button buttonPrevious = (Button) view.findViewById(R.id.buttonPrevious);
        final Button buttonNext = (Button) view.findViewById(R.id.buttonNext);

        buttonPrevious.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                psb.switchPrevious();
            }
        });

        buttonPrevious.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View view) {
                final Timer holdTimer = new Timer();
                holdTimer.scheduleAtFixedRate(new TimerTask() {
                    @Override
                    public void run() {
                        if (buttonPrevious.isPressed()) {
                            psb.setProgress(Math.max(psb.getProgress() - 5000, 0));
                        } else {
                            holdTimer.cancel();
                        }
                    }
                }, 1000, 500);
                return true;
            }
        });

        buttonNext.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                psb.switchNext();
            }
        });

        buttonNext.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View view) {
                final Timer holdTimer = new Timer();
                holdTimer.scheduleAtFixedRate(new TimerTask() {
                    @Override
                    public void run() {
                        if (buttonNext.isPressed()) {
                            psb.setProgress(Math.min(psb.getProgress() + 5000,
                                    psb.getCurrentPlayable().getMetadata().getLength()));
                        } else {
                            holdTimer.cancel();
                        }
                    }
                }, 1000, 500);
                return true;
            }
        });

        view.findViewById(R.id.buttonTogglePlay).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                psb.togglePlay();
            }
        });

        if(psb == null || psb.isEmpty()) {
            view.setVisibility(View.GONE);
        } else {
            onProgressChanged(psb.getProgress());
            onPlayableChanged(psb.getCurrentPlayable());
            onPlayStatusChanged(psb.isPlaying());
        }

        NoteStream.registerControlListener(this);
        NoteStream.registerProgressListener(this);
        NoteStream.registerPlayerServiceListener(new NoteStream.PlayerServiceListener() {
            @Override
            public void onPlayerServiceConnected(PlayerService.PlayerServiceBinder psb) {
                FragmentBarPlayer.this.psb = psb;
                if (!psb.isEmpty() && psb.getCurrentPlayable() != null) {
                    FragmentBarPlayer.this.onProgressChanged(psb.getProgress());
                    FragmentBarPlayer.this.onPlayableChanged(psb.getCurrentPlayable());
                    FragmentBarPlayer.this.onPlayStatusChanged(psb.isPlaying());
                    view.setVisibility(View.VISIBLE);
                }
            }

            @Override
            public void onPlayerServiceDisconnected() {
                view.setVisibility(View.GONE);
            }
        });
        return view;
    }

    @Override
    public void onProgressChanged(int progress) {
        ((ProgressBar) view.findViewById(R.id.songProgressBar)).setProgress(progress);
    }

    @Override
    public void onPlayableChanged(@NonNull Playable current) {
        ((TextView) view.findViewById(R.id.labelSongTitle)).setText(current.getMetadata().getTitle());
        ((TextView) view.findViewById(R.id.labelSongAuthor)).setText(current.getMetadata().getAuthor());
        ((ProgressBar) view.findViewById(R.id.songProgressBar)).setMax(current.getMetadata().getLength());
        onProgressChanged(0);
        view.setVisibility(View.VISIBLE);
    }

    @Override
    public void onPlayStatusChanged(boolean playing) {
        if (!playing) {
            view.findViewById(R.id.buttonTogglePlay).setBackground(ContextCompat.getDrawable(view.getContext(), R.drawable.ic_play_circle));
        } else {
            view.findViewById(R.id.buttonTogglePlay).setBackground(ContextCompat.getDrawable(view.getContext(), R.drawable.ic_pause_circle));
        }
    }

    @Override
    public void onShuffleStateChanged(boolean currentState) {

    }

    @Override
    public void onRepeatStateChanged(RepeatState currentState) {

    }
}

File: app/src/main/java/hr/caellian/notestream/gui/fragments/FragmentBarPlayer.kt
/*
 * Copyright (C) 2018 Tin Svagelj
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
 * persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
 * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
 * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package hr.caellian.notestream.gui.fragments

import android.app.Fragment
import android.content.Intent
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import hr.caellian.notestream.NoteStream
import hr.caellian.notestream.R
import hr.caellian.notestream.data.PlayerService
import hr.caellian.notestream.data.playable.Playable
import hr.caellian.notestream.gui.ActivityPlayer
import hr.caellian.notestream.util.RepeatState
import java.util.*

class FragmentBarPlayer : Fragment(), Playable.ControlListener, Playable.ProgressListener {
    internal var psb = NoteStream.instance?.psb

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_bar_player, container, false)

        view.setOnClickListener {
            val intent = Intent(view.context, ActivityPlayer::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT)
            startActivity(intent)
        }

        val buttonPrevious = view.findViewById<Button>(R.id.buttonPrevious)
        val buttonNext = view.findViewById<Button>(R.id.buttonNext)

        buttonPrevious.setOnClickListener { psb?.switchPrevious() }

        buttonPrevious.setOnLongClickListener {
            val holdTimer = Timer()
            holdTimer.scheduleAtFixedRate(object : TimerTask() {
                override fun run() {
                    if (buttonPrevious.isPressed) {
                        psb?.progress = Math.max(psb!!.progress - 5000, 0)
                    } else {
                        holdTimer.cancel()
                    }
                }
            }, 1000, 500)
            true
        }

        buttonNext.setOnClickListener { psb?.switchNext() }

        buttonNext.setOnLongClickListener {
            val holdTimer = Timer()
            holdTimer.scheduleAtFixedRate(object : TimerTask() {
                override fun run() {
                    if (buttonNext.isPressed) {
                        psb!!.progress = Math.min(psb!!.progress + 5000,
                                psb!!.currentPlayable!!.info.length)
                    } else {
                        holdTimer.cancel()
                    }
                }
            }, 1000, 500)
            true
        }

        view.findViewById<Button>(R.id.buttonTogglePlay).setOnClickListener { psb?.togglePlay() }

        if (psb?.isEmpty != true || psb?.isPlaying != true) {
            view.visibility = View.GONE
        }

        NoteStream.registerControlListener(this)
        NoteStream.registerProgressListener(this)
        NoteStream.registerPlayerServiceListener(object : NoteStream.PlayerServiceListener() {
            override fun onPlayerServiceConnected(psb: PlayerService.PlayerServiceBinder) {
                this@FragmentBarPlayer.psb = psb
                if (!psb.isEmpty && psb.isPlaying) {
                    this@FragmentBarPlayer.onProgressChanged(psb.progress)
                    this@FragmentBarPlayer.onPlayableChanged(psb.currentPlayable)
                    this@FragmentBarPlayer.onPlayStatusChanged(psb.isPlaying)
                    view.visibility = View.VISIBLE
                }
            }

            override fun onPlayerServiceDisconnected() {
                view.visibility = View.GONE
            }
        })
        return view
    }

    override fun onResume() {
        super.onResume()
        NoteStream.CONTROL_LISTENERS += this
        NoteStream.PROGRESS_LISTENERS += this
    }

    override fun onPause() {
        super.onPause()
        NoteStream.CONTROL_LISTENERS -= this
        NoteStream.PROGRESS_LISTENERS -= this
    }

    override fun onProgressChanged(progress: Int) {
        view?.findViewById<ProgressBar>(R.id.songProgressBar)?.progress = progress
    }

    override fun onPlayableChanged(current: Playable?) {
        view?.findViewById<TextView>(R.id.labelSongTitle)?.text = current?.title
        view?.findViewById<TextView>(R.id.labelSongAuthor)?.text = current?.author
        view?.findViewById<ProgressBar>(R.id.songProgressBar)?.max = current?.info?.length ?: 0
        onProgressChanged(0)
        view?.visibility = View.VISIBLE
    }

    override fun onPlayStatusChanged(playing: Boolean) {
        if (!playing) {
            view?.findViewById<Button>(R.id.buttonTogglePlay)?.background = ContextCompat.getDrawable(view.context, R.drawable.ic_play_circle)
        } else {
            view?.findViewById<Button>(R.id.buttonTogglePlay)?.background = ContextCompat.getDrawable(view.context, R.drawable.ic_pause_circle)
        }
    }

    override fun onShuffleStateChanged(currentState: Boolean) {}
    override fun onRepeatStateChanged(currentState: RepeatState) {}
}
